### PR TITLE
feat(CMSIS,Examples): Add C++ Support, Add Basic C++ Example, Consolidate GCC Makefiles

### PR DESCRIPTION
**Pull request by Jake-Carter**
This PR adds official C++ support for all micros.  The main change to support this feature was the modification of the linkerfiles to include some additional sections specific to C++.  These were copied from the mbed-os linkerfiles.  (See https://github.com/Analog-Devices-MSDK/msdk/commit/de043d594fca077b41ada8e4339df3a7871c57d2).  Project Makefiles now include .cpp files in auto-search.  

Additionally, the gcc/gcc_riscv.mk files have been consolidated to a single location (`Libraries/CMSIS/Device/Maxim/GCC`) for all micros.  This improves our ability to apply core changes universally.

All micros now include a `Hello_World_Cpp` example demonstrating basic class creation and instancing.

### CODE REVIEW

File: Examples/MAX32520/Hello_World_Cpp/main.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    my_class.hello_method();
+
+    printf("Hello World!\n");
+
+    while (1) {
+        LED_On(LED1);
+        MXC_Delay(500000);
+        LED_Off(LED1);
+        MXC_Delay(500000);
+        printf("count : %d\n", my_class.get_count());
```
Comment (ozersa) [inline Examples/MAX32520/Hello_World_Cpp/main.cpp:93]: It will be better to use std::cout instead of printf.

Comment (Jake-Carter) [reply Examples/MAX32520/Hello_World_Cpp/main.cpp:93]: It seems that iostream increase the code size by about 293K

Before...
```shell
arm-none-eabi-size --format=berkeley /home/jakecarter/repos/fork/msdk/Examples/MAX78000/Hello_World_Cpp/build/max78000.elf
   text    data     bss     dec     hex filename
  36124    2504    1228   39856    9bb0 /home/jakecarter/repos/fork/msdk/Examples/MAX78000/Hello_World_Cpp/build/max78000.elf
```

```diff
--- a/Examples/MAX78000/Hello_World_Cpp/main.cpp
+++ b/Examples/MAX78000/Hello_World_Cpp/main.cpp
@@ -45,6 +45,7 @@
 #include "led.h"
 #include "board.h"
 #include "mxc_delay.h"
+#include <iostream>

 /***** Definitions *****/

@@ -83,7 +84,7 @@ int main()
     HelloClass my_class = HelloClass();
     my_class.hello_method();

-    printf("Hello World!\n");
+    std::cout << "Hello world!\n";

     while (1) {
         LED_On(LED1);
```

after...
```shell
arm-none-eabi-size --format=berkeley /home/jakecarter/repos/fork/msdk/Examples/MAX78000/Hello_World_Cpp/build/max78000.elf
   text    data     bss     dec     hex filename
 329236    2788    7000  339024   52c50 /home/jakecarter/repos/fork/msdk/Examples/MAX78000/Hello_World_Cpp/build/max78000.elf
```

File: Examples/MAX32520/Hello_World_Cpp/main.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+#include "board.h"
+#include "mxc_delay.h"
+
+/***** Definitions *****/
+
+/***** Globals *****/
+
+/***** Functions *****/
+
+class HelloClass {
```
[the remark this answers is not in the document]
Comment (Jake-Carter) [reply Examples/MAX32520/Hello_World_Cpp/main.cpp]: Thanks, this is a better example.  I applied it in https://github.com/Analog-Devices-MSDK/msdk/pull/619/commits/f0edb97154051b8f0731c8e6588ee0195fc2344b

**Review by sihyung-maxim — APPROVED** [review]
Fancy, looks good.

Just writing a note for the future, but we can reference how mbed sets up their libraries since it's in C++ when we dive deeper into implementing C++ drivers into the SDK. Many companies and customers use mbed so might be a good place to start.

File: Examples/MAX32520/Hello_World_Cpp/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+## Software
+
+### Project Usage
+
+Universal instructions on building, flashing, and debugging this project can be found in the **[MSDK User Guide](https://analog-devices-msdk.github.io/msdk/USERGUIDE/)**.
+
+### Project-Specific Build Notes
+
+* This project comes pre-configured for the MAX78000EVKIT.  See [Board Support Packages](https://analog-devices-msdk.github.io/msdk/USERGUIDE/#board-support-packages) in the UG for instructions on changing the target board.
```
Comment (Jacob-Scheiffler) [inline Examples/MAX32520/Hello_World_Cpp/README.md:15]: Looks like the Project-Specific Build Notes got copied into each README

File: Examples/MAX32520/Hello_World_Cpp/main.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+/***** Globals *****/
+
+/***** Functions *****/
+
+class LED {
+public:
+    explicit LED(int index)
+    {
+        idx = index;
```
Comment (Jacob-Scheiffler) [inline Examples/MAX32520/Hello_World_Cpp/main.cpp:59]: Consider checking for a valid index here.

**Review by Jacob-Scheiffler — CHANGES_REQUESTED** [review]

**Comment by Jacob-Scheiffler** [comment]
Is Libraries/CMSIS/Device/Maxim/GCC a valid CMSIS directory? Not saying it's wrong, just asking 'cause I don't know the answer and was under the impression that CMSIS had a strict format for their directories.

File: Examples/MAX32520/Hello_World_Cpp/README.md (hunk 3 above)
Comment (Jake-Carter) [reply Examples/MAX32520/Hello_World_Cpp/README.md:15]: Thanks, good catch

Comment (Jake-Carter) [reply Examples/MAX32520/Hello_World_Cpp/README.md]: https://github.com/Analog-Devices-MSDK/msdk/pull/619/commits/50d5e14fe13ef79420406d963373257d9a647584

**Comment by Jake-Carter** [comment]
> Is Libraries/CMSIS/Device/Maxim/GCC a valid CMSIS directory? Not saying it's wrong, just asking 'cause I don't know the answer and was under the impression that CMSIS had a strict format for their directories.

I don't think there are any hard rules.  There are [core device templates](https://arm-software.github.io/CMSIS_5/Core/html/templates_pg.html) that we've followed.  I think we get some leeway since we're already inside of the "Maxim" sub-folder

**Review by Jacob-Scheiffler — APPROVED** [review]

File: Examples/MAX32690/Hello_World/README.md [hunk 5]
```diff
@@ -27,11 +27,14 @@ Universal instructions on building, flashing, and debugging this project can be
 The Console UART of the device will output these messages:
 
 ```
+Hello class!
+Hello method!
 Hello World!
```
Comment (lorne-maxim) [inline Examples/MAX32690/Hello_World/README.md:32]: Is this expected output correct for the non-CPP example?

File: Libraries/CMSIS/Device/Maxim/MAX32572/Source/GCC/gcc.mk [hunk 6]
```diff
@@ -34,407 +34,11 @@
...
-clean:
-	@rm -rf ${BUILD_DIR} ${wildcard *~}
-
-${BUILD_DIR}/${PROJECT}.elf: ${LIBS} ${OBJS} ${LINKERFILE}
-${BUILD_DIR}/${PROJECT}.a: ${OBJS}
-
-# Create a goal to exercise the library build dependencies
-.PHONY: FORCE
-FORCE:
+OLD_GCC_DIR := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
```
Comment (lorne-maxim) [inline Libraries/CMSIS/Device/Maxim/MAX32572/Source/GCC/gcc.mk:37]: This part hasn't been released yet so is support for old style projects necessary?

File: CPPLINT.cfg [hunk 7]
```diff
@@ -5,6 +5,7 @@ filter=-readability/casting
 filter=-whitespace/braces
 filter=-whitespace/comments
 filter=-whitespace/line_length
+filter=-whitespace/indent
```
[the remark this answers is not in the document]
Comment (Jake-Carter) [reply CPPLINT.cfg:9]: Yes, there was a conflict between cpplint and clang-format but only for C++ code.

cpplint wanted
```C
class MyClass {
  public:
  // ...
}
```

where clang-format wanted
```C
class MyClass {
public:
// ...
}
```

[the remark this answers is not in the document]
Comment (lorne-maxim) [reply CPPLINT.cfg:9]: Well, clearly cpplint is superior then, lol!

File: Examples/MAX32690/Hello_World/README.md (hunk 5 above)
Comment (Jake-Carter) [reply Examples/MAX32690/Hello_World/README.md]: Thanks, fixed in https://github.com/Analog-Devices-MSDK/msdk/pull/619/commits/9e9a91c38b07aa42cbce6d9fc72081a856c422ae

File: Libraries/CMSIS/Device/Maxim/MAX32572/Source/GCC/gcc.mk (hunk 6 above)
Comment (Jake-Carter) [reply Libraries/CMSIS/Device/Maxim/MAX32572/Source/GCC/gcc.mk]: Good point, I deleted the files in https://github.com/Analog-Devices-MSDK/msdk/pull/619/commits/e929295ea899f24c74efca7a1f9438b0db6b2235.  I also updated all the parts to point direct to the new locations.

**Pull request merged**